compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -326,6 +328,17 @@ open class ClassCodegen protected constructor(
         }
     }
 
+    // JVMS 4.7.2, The ConstantValue Attribute: https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.2
+    // Constant-initialized primitive static final fields act like language constants: they are inlined and can be used
+    // in annotations and Java switch cases.
+    fun fieldConstantValue(field: IrField): IrConst<*>? {
+        // TODO return null if NoConstantValueAttributeForNonConstVals is set and the property is not explicitly const
+        val asmType = typeMapper.mapType(field.descriptor)
+        if (!field.isStatic || !field.isFinal || (!AsmUtil.isPrimitive(asmType) && asmType != AsmTypes.JAVA_STRING_TYPE))
```

Comment:
Since Kotlin 1.4 (language feature: NoConstantValueAttributeForNonConstVals), we generate ConstantValue attributes for `const val`s only. That's because such fields can be used in constant expressions (e.g., in annotations) in Java code, and if someone changes property definition, it can implicitly break binary compatibility.

Reply:
Yeah, that's what the TODO is about. I'm trying to figure out how to correctly determine constness without using descriptors right now.

Reply:
irField.corresponginPropertySymbol.owner.isConst?

Reply:
That's unfortunately not always correct for Java fields.

Reply:
I've added a condition that almost does the right thing; the only exception is an inherited Java field, in which case the field has origin FAKE_OVERRIDE with empty overriddenSymbols, so I see no way to check that it actually came from Java.